torchtune/dev/grpo/types.py

```diff
@@ -5,9 +5,203 @@
+            policy_version=self.policy_version
+        )
+
+@dataclass
+class PackedTrajectory:
+    """
+    Represents a batch of trajectories packed into single tensors for efficient
+    model processing.
+
+    Attributes:
+        tokens (torch.Tensor): Concatenated prompt and response tokens for all sequences
+            in the batch, potentially padded. Shape: (packed_seq_len,).
+            Example: [P1, R1, P2, R2, ...].
+        attention_mask (torch.Tensor): Block-diagonal causal attention mask for the
+            packed sequences. Shape: (packed_seq_len, packed_seq_len).
+        position_ids (torch.Tensor): Position IDs for the packed sequences. Shape: (packed_seq_len,).
+        response_mask (torch.Tensor): Boolean mask indicating the positions of response
+            tokens within the `tokens` tensor. Shape: (packed_seq_len,).
+        sequence_mask (torch.Tensor): Integer mask indicating which sequence each token
+            belongs to. Shape: (packed_seq_len,).
+        prompt_lens (torch.Tensor): Lengths of the prompts for each sequence in the batch.
+            Shape: (num_sequences,).
+        response_lens (torch.Tensor): Lengths of the responses for each sequence in the
+            batch. Shape: (num_sequences,).
+        sequence_map (torch.Tensor): Start and end indices for each sequence within the
+            `tokens` tensor. Shape: (num_sequences, 2).
+        packed_seq_len (int): The total length of the `tokens` tensor, including any
+            padding.
+        ref_logprobs (Optional[torch.Tensor]): Concatenated reference model log
+            probabilities for all response tokens. Shape: (total_response_tokens,).
+        advantages (Optional[torch.Tensor]): Concatenated advantage values for all
+            response tokens. Shape: (total_response_tokens,).
+        targets (Optional[torch.Tensor]): Concatenated target tokens (usually the
+            response tokens shifted). Shape: (total_response_tokens,).
+        sequence_ids (Optional[List[str]]): List of unique sequence identifiers
+            corresponding to the packed sequences.
+        group_ids (Optional[List[str]]): List of group identifiers corresponding to the
+            packed sequences.
+        actual_total_tokens (int): The total number of tokens in the batch, excluding
+            any padding tokens used in `tokens`.
+    """
+    tokens: torch.Tensor
+    attention_mask: torch.Tensor
+    position_ids: torch.Tensor
+    response_mask: torch.Tensor
+    sequence_mask: torch.Tensor
+    prompt_lens: torch.Tensor
+    response_lens: torch.Tensor
+    sequence_map: torch.Tensor
+    packed_seq_len: int
+    ref_logprobs: Optional[torch.Tensor] = None
+    advantages: Optional[torch.Tensor] = None
+    targets: Optional[torch.Tensor] = None
+    sequence_ids: Optional[List[str]] = None
+    group_ids: Optional[List[str]] = None
+    actual_total_tokens: int = 0
+
+@dataclass
+class GRPOStats:
```

Comment:
Can be made more generic: there's nothing inherently about GRPO as an algo compared to eg PPO or any other online RL algo. Maybe `TrainingStats`?